pytensor/link/pytorch/dispatch/nlinalg.py

```diff
@@ -0,0 +1,12 @@
+import torch
+
+from pytensor.link.pytorch.dispatch import pytorch_funcify
+from pytensor.tensor.math import Dot
+
+
+@pytorch_funcify.register(Dot)
+def pytorch_funcify_Dot(op, **kwargs):
+    def dot(x, y):
+        return torch.matmul(x, y)
```

Comment:
Yep. For PyTorch, [torch.dot](https://pytorch.org/docs/stable/generated/torch.dot.html) is only the inner product for 1D vectors.

Reply:
But matmul is not going to do the expected thing for 3d x, y. Numpy/PyTensor dot will result in an output that concatenates the batch dimensions, whereas matmul (in numpy/pytensor/pytorch) broadcasts the batch dimensions. The docs of JAX are quite succint in the difference between dot and matmul: 
![image](https://github.com/pymc-devs/pytensor/assets/28983449/9b691e6f-0b46-4085-8be1-02745ad71638)

We need to find a way to implement that behavior in torch. I've seen suggestions of using `torch.einsum` or view (so that torch.matmul then behaves the same as the equivalent dot

Reply:
Yeah I think `tensordot(a, b, axes=[(-1),(-2)]` is equivalent to `np.dot` when inputs >= 2 ndim (may need a different implementation when at least one input < 2 ndim

Reply:
I tweaked the backend to use `matmul` if one input is a vector or when both do not have batch dimension, and use `tensordot` otherwise. However, a "NotImplementedError` error keeps appearing in testing.
```bash
NotImplementedError: No PyTorch conversion for the given `Op`: Shape_i{2}.
Check out `[https://github.com/pymc-devs/pytensor/issues/821`](https://github.com/pymc-devs/pytensor/issues/821%60) for progress or to request we prioritize this operation
```
This `Shape_i` is one Op marked in the list. I think I will need to implement this `Op` first before pushing ahead?

Reply:
I guess the shape is being introduced somewhere in your test graph, you shouldn't need for the purpose of testing the Op. Can you dprint the whole fgraph by the time we start dispatching it into PyTensor?

Reply:
@ricardoV94 Yes the `shape_i` is introduced in rewrite. The output of this snippet
```py
import pytensor
from pytensor.tensor.math import dot
from pytensor.tensor.type import tensor3
import numpy as np

a, b = tensor3("a"), tensor3("b")
A = np.zeros((3, 2, 4))
B = np.zeros((3, 4, 1))
c = dot(a, b)
f = pytensor.function([a, b], c)
pytensor.dprint(f)
```
gives
```bash
Reshape{4} [id A] 13
 ├─ Dot22 [id B] 12
 │  ├─ Reshape{2} [id C] 11
 │  │  ├─ a [id D]
 │  │  └─ MakeVector{dtype='int64'} [id E] 10
 │  │     ├─ -1 [id F]
 │  │     └─ Shape_i{2} [id G] 9
 │  │        └─ a [id D]
 │  └─ Reshape{2} [id H] 8
 │     ├─ Transpose{axes=[1, 0, 2]} [id I] 7
 │     │  └─ b [id J]
 │     └─ MakeVector{dtype='int64'} [id K] 6
 │        ├─ Shape_i{1} [id L] 5
 │        │  └─ b [id J]
 │        └─ -1 [id F]
 └─ MakeVector{dtype='int64'} [id M] 4
    ├─ Shape_i{0} [id N] 3
    │  └─ a [id D]
    ├─ Shape_i{1} [id O] 2
    │  └─ a [id D]
    ├─ Shape_i{0} [id P] 1
    │  └─ b [id J]
    └─ Shape_i{2} [id Q] 0
       └─ b [id J]
```
for `pytensor.dprint(c)` it is
```bash
Reshape{4} [id A]
 ├─ dot [id B]
 │  ├─ Reshape{2} [id C]
 │  │  ├─ DimShuffle{order=[0,1,2]} [id D]
 │  │  │  └─ a [id E]
 │  │  └─ MakeVector{dtype='int64'} [id F]
 │  │     ├─ Cast{int64} [id G]
 │  │     │  └─ -1 [id H]
 │  │     └─ Mul [id I]
 │  │        ├─ 1 [id J]
 │  │        └─ Subtensor{i} [id K]
 │  │           ├─ Shape [id L]
 │  │           │  └─ a [id E]
 │  │           └─ 2 [id M]
 │  └─ Reshape{2} [id N]
 │     ├─ Transpose{axes=[1, 0, 2]} [id O]
 │     │  └─ b [id P]
 │     └─ MakeVector{dtype='int64'} [id Q]
 │        ├─ Mul [id R]
 │        │  ├─ 1 [id S]
 │        │  └─ Subtensor{i} [id T]
 │        │     ├─ Shape [id U]
 │        │     │  └─ b [id P]
 │        │     └─ 1 [id V]
 │        └─ Cast{int64} [id W]
 │           └─ -1 [id X]
 └─ MakeVector{dtype='int64'} [id Y]
    ├─ Subtensor{i} [id Z]
    │  ├─ Shape [id L]
    │  │  └─ ···
    │  └─ 0 [id BA]
    ├─ Subtensor{i} [id BB]
    │  ├─ Shape [id L]
    │  │  └─ ···
    │  └─ 1 [id BC]
    ├─ Subtensor{i} [id BD]
    │  ├─ Shape [id U]
    │  │  └─ ···
    │  └─ 0 [id BE]
    └─ Subtensor{i} [id BF]
       ├─ Shape [id U]
       │  └─ ···
       └─ 2 [id BG]
```

Reply:
> I guess the shape is being introduced somewhere in your test graph during rewrites. Can you dprint the whole fgraph by the time we start dispatching it into PyTensor?
> 
> We can try and exclude those rewrites, which is important anyway to make sure we are actually testing the dispatch

How exactly can we disable the rewrites in a function?

Reply:
You pass a mode. Something like `mode = pytorch_mode.excluding("BlasOpt"). But these should be excluded from the PyTorch mode already. Can you put `pytensor.config.optimizer_verbose=True` to see which rewrite is introducing dot22?

Reply:
The first one is not the graph you get with torch though, because we exclude BlasOpt rewrites. However the initial call to `dot` already does something more for the 3d case. I didn't know that

Reply:
Ah okay `dot` actually only works for 1 or 2d inputs. Above that calling `dot` triggers pytensor `tensordot` which does the reshape logic to reduce it back to a dot.

https://github.com/pymc-devs/pytensor/blob/ca102983507d0eb3fd53f532a55290eb22bd2483/pytensor/tensor/math.py#L1829-L1855

https://github.com/ricardoV94/pytensor/blob/ca102983507d0eb3fd53f532a55290eb22bd2483/pytensor/tensor/math.py#L2021-L2033

So that's nice it means we can actually just use `matmul`/`dot` in our dispatch implentation because there will never be a node with ndim > 2. For testing purposes we don't need to cover > 2d then, because those are just reshape + 2d dot. Once we have this and reshape they should work out of the box.

Reply:
@ricardoV94 Perfect! I should have checked that. Anyway, I reverted the logic back to the original.